File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.py
import sys
from collections import deque

for test_case in range(1):
    n, k = map(int, sys.stdin.readline().split())

    dp = [float("inf") for i in range(100001)]

    q = deque()

    q.append((n, 0))
    dp[n] = 0
    while q:
        cur_pos, cur_cnt = q.popleft()

        cases = [(cur_pos - 1, 1), (cur_pos + 1, 1), (cur_pos * 2, 0)]
        if dp[cur_pos] < cur_cnt:
            continue

        if cur_pos == k:
            print(cur_cnt)
            break

        for new_pos, time in cases:
            if 0 <= new_pos <= 100000 and dp[new_pos] > cur_cnt + time:
                dp[new_pos] = cur_cnt + time
                q.append((new_pos, cur_cnt + time))

File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.java
import java.util.*;
import java.io.*;

public class Main
{

	public static void main(String args[]) throws Exception
	{
//		BufferedReader br = new BufferedReader(new FileReader("./input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));	
		
		String[] str = br.readLine().split(" ");
		int n = Integer.parseInt(str[0]);
		int k = Integer.parseInt(str[1]);
		
		int[] dp = new int[100001];
		
		for (int i = 0; i < 100001; i ++) dp[i] = Integer.MAX_VALUE;
		
		Deque<int[]> q = new LinkedList<>();
		int[] tmp = {n, 0};
		q.add(tmp);
		
		while (!q.isEmpty()) {
			tmp = q.pollFirst();
			int cur_node = tmp[0];
			int cur_cnt = tmp[1];
			
			int[][] cases = {{cur_node - 1, 1}, {cur_node + 1, 1}, {cur_node * 2, 0}};
			if (dp[cur_node] < cur_cnt ) continue;
			if (cur_node == k) {
				System.out.println(cur_cnt);
				break;
			}
			
			for (int[] c : cases) {
				int new_node = c[0];
				int new_cnt = c[1];
				
				if (0 <= new_node && new_node <= 100000 && dp[new_node] > cur_cnt + new_cnt) {
					dp[new_node] = cur_cnt + new_cnt;
					int[] t = {new_node, cur_cnt + new_cnt};
					q.add(t);
				}
			}
			
		}
		
		
    }	
		
}